File: get_data/ins_mortgage_rate.jl
using HTTP
using Gumbo
using Cascadia
using DataFrames
using LibPQ
include(joinpath(pwd(), "postgresql_ops.jl"))
using .PostgresqlOps

response = HTTP.get("https://www.interest.co.nz/borrowing")
html = parsehtml(String(response.body))

h2_nodes = eachmatch(Selector("h2"), html.root)
banks = first(n for n in h2_nodes if strip(Gumbo.text(n)) == "Banks")
banks_parent = banks.parent
data_table = Cascadia.matchFirst(Selector("table#interest_financial_datatable"), banks_parent)

headers = Cascadia.matchFirst(Selector("thead tr"), data_table)
col_names = [strip(nodeText(col)) for col in eachmatch(Selector("th"), headers)]
n_cols = length(col_names)
rows = eachmatch(Selector("tbody tr"), data_table)
rows_1 = Vector{Dict{String,String}}()
current_bank = ""
for row in rows
    global current_bank
    cols = eachmatch(Selector("td"), row)
    n_cols_this_row = length(cols)

    # Detect bank name from primary_row (img title or link text)
    inst_cell = cols[1]
    img_nodes = eachmatch(Selector("img"), inst_cell)
    a_nodes = eachmatch(Selector("a"), inst_cell)
    if !isempty(img_nodes)
        current_bank = img_nodes[1].attributes["title"]
    elseif !isempty(a_nodes)
        current_bank = a_nodes[1].attributes["title"]
    end

    # Remove ads
    current_bank = replace(current_bank, "Home Loans %u2013 Apply now or find out more" => "")  # TSB
    current_bank = replace(current_bank, r"click.*contact" => "")  # SBS, BNZ
    current_bank = replace(current_bank, r" - .*" => "")  # ASB
    current_bank = strip(current_bank)
    if length(current_bank) > 32
        current_bank = current_bank[1:32]
    end
 
    if n_cols_this_row == n_cols  # Normal row
        rd = Dict{String,String}()
        for (i, col) in enumerate(cols)
            val = strip(nodeText(col))
            if col_names[i] == "Product" && length(val) > 32
                val = val[1:32]
            end
            rd[col_names[i]] = val
        end
        rd["Institution"] = current_bank  # Override with cleaned bank name
        push!(rows_1, rd)
    else  # Colspan row — merge 18-month value into previous row
        special_td = cols[3]
        value_spans = eachmatch(Selector(".interest_data-subcell-special_line_value"), special_td)
        if !isempty(value_spans) && !isempty(rows_1)
            rows_1[end]["18months"] = strip(nodeText(value_spans[1]))
        end
    end
end

parse_rate(s::AbstractString) = isempty(s) ? missing : parse(Float64, s)

col_rename = Dict(
    "Variable floating" => :Floating,
    "6months" => :_6_Months,
    "1year" => :_1_Year,
    "18months" => :_18_Months,
    "2years" => :_2_Years,
    "3years" => :_3_Years,
    "4years" => :_4_Years,
    "5years" => :_5_Years,
)

df = DataFrame(
    Bank = [r["Institution"] for r in rows_1],
    Product = [r["Product"] for r in rows_1],
)
for (src, dst) in col_rename
    df[!, dst] = [parse_rate(get(r, src, "")) for r in rows_1]
end

data_cols = collect(values(col_rename))
filter!(r -> any(!ismissing, [r[c] for c in data_cols]), df)
unique!(df, [:Bank, :Product])

const c = LibPQ.Connection(ENV["NEON_DB"])
insert_if_not_exists(c, df, ["Date", "Bank", "Product"], "ins_mortgage_rate")
close(c)

File: get_data/ins_mortgage_rate.py
import logging
import os
import re
import sys
from io import StringIO

import pandas as pd
from bs4 import BeautifulSoup
from requests import Session
from sqlalchemy import create_engine

from postgresql_upsert import insert_if_not_exists

# %% Initialization.
logging.basicConfig(
    level=logging.INFO,
    format="[%(levelname)s] %(message)s",
    handlers=[
        logging.StreamHandler(sys.stdout),
    ],
)
session = Session()

# %% Fetch borrowing page.
response = session.get("https://www.interest.co.nz/borrowing")
response.raise_for_status()
html = BeautifulSoup(response.text, "html.parser")
banks_h2 = html.find("h2", string=re.compile("Banks"))
data_table_raw = banks_h2.parent.find(
    "table", {"id": "interest_financial_datatable"})

# %% Get bank name.
rows = data_table_raw.find("tbody").find_all("tr")
bank_names = []
for row in rows:
    bank_col = row.find("td")  # Assume first column is bank logo
    container = bank_col.find("img") or bank_col.find("a") or bank_col
    bank_name = container.get('title') or container.text or pd.NA
    # Remove ads or suffixes
    if isinstance(bank_name, str):
        bank_name = re.sub(r"Home.*more", "", bank_name)  # TSB
        bank_name = re.sub(r"click.*contact", "", bank_name)  # SBS, BNZ
        bank_name = re.sub(r" - .*", "", bank_name)  # ASB
        bank_name = bank_name.strip()[:32]
    bank_names.append(bank_name)

# %% Get column names.
data_table = pd.read_html(StringIO(str(data_table_raw)))[0]
assert len(bank_names) == data_table.shape[0]
data_table['bank'] = bank_names
data_table['bank'] = data_table['bank'].ffill()

_18_months_rows = data_table.apply(
    lambda col: col.str.contains(r"18 months =", na=False, regex=True)
).any(axis=1)

table_1 = data_table.iloc[~_18_months_rows, :]
table_2 = data_table.iloc[_18_months_rows, :]
table_1['_18_months'] = (table_2.iloc[:, 2].str.extract(r'=\s*([0-9]*\.?[0-9]+)')
                         .astype(float))
table_1.rename(columns={
    "Product": "product",
    "Variable floating": "floating",
    "6 months": "_6_months",
    "1 year": "_1_year",
    "2 years": "_2_years",
    "3 years": "_3_years",
    "4 years": "_4_years",
    "5 years": "_5_years",
}, inplace=True)
del table_1['Institution']

# %% Export.
data_cols = table_1.columns.difference(['bank', 'product'])
table_1.dropna(subset=data_cols, how="all", inplace=True)
table_1.drop_duplicates(subset=["bank", "product"], inplace=True)
table_1['date'] = pd.Timestamp('now', tz='Pacific/Auckland').date()
table_1 = table_1.convert_dtypes()

engine = create_engine(os.environ["NEON_DB"])
insert_if_not_exists(
    engine, table_1,
    ["date", "bank", "product"],
    "ins_mortgage_rate"
)
